node/test/chan_drain.js

```diff
@@ -0,0 +1,374 @@
+    }
+
+    function sendOne(client, _, done) {
+        reqN++;
+        assert.comment('sending request ' + reqN);
+        client.request().send('echo', 'such', 'mess' + reqN, done);
+    }
+
+    function drained() {
+        assert.pass('drain happened');
+        server.close(closed);
+    }
+
+    function closed(err) {
+        if (err) {
+            finish(err);
+            return;
+        }
+
+        assert.pass('server closed');
+        setTimeout(finish, 1);
+    }
+
+    function finish(err) {
+        finishCount = checkFinish(assert, err, cluster, finishCount);
+    }
+});
+
+// TODO: test draining of outgoing reqs
+// TODO: currently the exempt test provokes mismatched onReqDone callback warn logs
+
+function setupTestClients(cluster, services, callback) {
+    var i;
+    var clients = {};
+    var serverRoot = cluster.channels[0];
+
+    for (i = 0; i < services.length; i++) {
+        var service = services[i];
+        var clis = clients[service] = [];
+        for (var j = 1; j < cluster.channels.length; j++) {
+            var client = setupServiceClient(cluster.channels[j], service);
+            clis.push(client);
+            client.peers.add(serverRoot.hostPort);
+        }
+    }
+
+    collectParallel(cluster.channels.slice(1), idEach, ided);
+
+    function idEach(chan, i, done) {
+        var peer = chan.peers.add(serverRoot.hostPort);
+        peer.waitForIdentified(done);
+    }
+
+    function ided(err, res) {
+        for (var i = 0; i < res.length; i++) {
+            if (res[i].err) {
+                callback(err, clients);
+                return;
+            }
+        }
+        callback(null, clients);
+    }
+}
+
+function setupServiceServer(rootchan, service, delay) {
+    var chan = rootchan.makeSubChannel({
+        serviceName: service
+    });
+    chan.register('echo', echo);
+
+    return chan;
+
+    function echo(req, res, arg2, arg3) {
+        req.channel.timers.setTimeout(respond, delay);
+
+        function respond() {
+            res.headers.as = 'raw';
+            res.send(arg2, arg3);
+        }
+    }
+}
+
+function setupServiceClient(rootchan, service) {
+    return rootchan.makeSubChannel({
+        serviceName: service,
+        requestDefaults: {
+            timeout: 20,
+            hasNoParent: true,
+            retryflags: {
+                never: true
+            },
+            serviceName: service,
+            headers: {
+                as: 'raw',
+                cn: service + 'Client'
+            }
+        }
+    });
+}
+
+function checkSendsDone(name, check, callback) {
+    return sendsDone;
+
+    function sendsDone(err, res) {
+        if (err) {
+            callback(err);
+            return;
+        }
+        for (var i = 0; i < res.length; i++) {
+            var desc = name + ' res[' + i + ']: ';
+            check(desc, res[i], i);
+        }
+        callback();
+    }
+}
+
+function checkFinish(assert, err, cluster, finishCount) {
```

Comment:
This might be nicer as `assert.once('end', checkFinish);` that just does the assertCleanState() stuff.